### PR TITLE
Enhancement#5 add threshold checker

### LINKED ISSUE #5: Add an option of "threshold" to the early-stopping mechanism
The suggestion to add an "epsilon threshold" to the early-stopping callback can be implemented as follows:
- add a class named "thresholdChecker" as part of monitor callback
- insert various criteria to this class (such as absolute threshold or relative threshold)
- implement a function in this class to check if criteria is met
- invoke this function in monitor callback before changing early_stopping variable

### PR DESCRIPTION
**Pull request by NaorHaba**
- added the threshold checker with relevant tests
- fixed pipfile to include any version of numpy

resolved #5

### CODE REVIEW

File: lpd/utils/threshold_checker.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    Check if the current value is better than the previous best value according to different threshold criteria
+    This is an abstract class meant to be inherited by different threshold checkers
+    Can also be inherited by the user to create custom threshold checkers
+    """
+    def __init__(self, threshold: float):
+        self.threshold = threshold
+
+    @abstractmethod
+    def is_new_value_higher(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
```
Comment (RoySadaka) [inline lpd/utils/threshold_checker.py]: is_new_value_higher & is_new_value_lower functions are not "structured",
meaning, the main goal of lpd is that you know how everything behaves by 
reading the trainer.py

Look at CallbackMonitor for example, you define its behaviour in the ctor, and know
how it behaves.

i think we should align with it by defining the ThresholdChecker with the supported direction (higher/lower)
maybe create a new enum ? 
ThresholdCheckerMode.Higher
ThresholdCheckerMode.Lower

that way the ThresholdChecker will also only implement the default \_\_call\_\_ function
to check if condition is met or not

File: lpd/utils/threshold_checker.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def __init__(self, threshold: float = 0.0):
+        if threshold < 0:
+            raise ValueError(f"Threshold must be non-negative, but got {threshold}")
+        super(RelativeThresholdChecker, self).__init__(threshold)
+
+    def is_new_value_higher(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return (new_value - old_value) / old_value > self.threshold
+
+    def is_new_value_lower(self, new_value: Union[float, Tensor], old_value: Union[float, Tensor]) -> bool:
+        return (old_value - new_value) / old_value > self.threshold
```
Comment (RoySadaka) [inline lpd/utils/threshold_checker.py]: looks like a bug
should be 
(old_value - new_value) / new_value
say 
threshold = 0.05 (5%)
old_value = 10
new_value = 9.51
a drop of 5% from 10 will give 10/1.05 = ~9.52
yet, (old_value - new_value) / old_value = (10-9.51)/10 = 0.049
which is lower than 0.05, but the threshold should be met in this case

the fix, (old_value - new_value) / new_value will yield:
(10-9.51)/9.51 = 0.0515 which is greater than the threshold, as expected

Comment (NaorHaba) [reply lpd/utils/threshold_checker.py]: So if I understand correctly, you are saying that the 5% change should be according to the new value?
This isn't how I was thinking about this method. The way I see it, if you lose 5% out of 10 you get 9.5. This means that in order to pass a 5% threshold you need to "lose" more value (in the is_new_value_lower case). So I believe that your example shouldn't meet the threshold requirement 😅

Comment (RoySadaka) [reply lpd/utils/threshold_checker.py]: got ya, sorry for the confusion

File: lpd/utils/threshold_checker.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     is greater than or equal to a given threshold
+
+    Args:
+        monitor_mode: MIN or MAX
+        threshold - the threshold to check (must be non-negative)
+    """
+    def __init__(self, monitor_mode: MonitorMode, threshold: float = 0.0):
+        if threshold < 0:
+            raise ValueError(f"Threshold must be non-negative, but got {threshold}")
+        super(AbsoluteThresholdChecker, self).__init__(threshold)
```
Comment (RoySadaka) [inline lpd/utils/threshold_checker.py:35]: add monitor_mode to super(...) as well

**Review by RoySadaka — APPROVED** [review]

**Pull request merged**